Add Anet ET4/ET4P motherboard & Anet TFT28/TFT35 support

thisiskeithb opened the pull request:
### Description

The Anet ET4 (with A4988s) and ET4P (with TMC2208s in standalone mode) are STM32F407VGT6-based motherboards used in Anet's ET4 and ET5 series of printers (standard, X, and Pro) as well as the [Labist ET4](https://labists.com/products/3d-printer-et4-diy-kit). This PR brings in initial support for these boards as well as Anet's TFT28 & TFT35 TFTs from [davidtgbe/Marlin](https://github.com/davidtgbe/Marlin) / [anet-3d/Marlin/tree/Anet-Marlin-bugfix-2.1.0](https://github.com/anet-3d/Marlin/tree/Anet-Marlin-bugfix-2.1.0).

### Benefits

Existing Anet ET4/5 users who flash their boards with the [OpenBLT bootloader](https://github.com/davidtgbe/openblt) will be able to use Marlin on their printers.

_Note: Anet plans to ship Marlin on their printers at some point in the future, so users won't have to update their bootloader._

### Configurations

[Any of the Anet ET4 or ET5 configs will work for testing](https://github.com/MarlinFirmware/Configurations/tree/import-2.0.x/config/examples/Anet).

thisiskeithb committed:
Initial Anet ET4 Board Support

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/20280

thinkyhead committed:
Cosmetic tweaks

thinkyhead committed:
Endian color patch

thinkyhead committed:
TFT FMSC patches

thinkyhead commented:
I've brought in a number of changes from the ET4 branch I was working on. I'm not about the Color and FSMC patches, but they're included here for evaluation.

thisiskeithb commented:
> I've brought in a number of changes from the ET4 branch I was working on.

Ok, great! I didn’t realize you had something going, so apologies if this was too early.

Anet reached out to me and sent along the referenced firmware so I wanted to get a draft going for feedback/testing.

> I'm not about the Color and FSMC patches, but they're included here for evaluation.

Good to know. I’ve seen stock ET4/5 machines running with `TFT_COLOR_UI`, but since I’m waiting on a real printer to arrive, I wasn’t able to test on real hardware yet.

thinkyhead commented:
> apologies if this was too early…

I have been meaning to get it tested, since I do have an ET4 here. The main question is whether the `IS_ANET_ET` customizations here (making the user interface blue and orange, etc.) are desirable or necessary. Also, the pins file I have specifies a `CUSTOM_CABLE` for the SD connection so we might need to look at that. Also, I have some differences in my `platformio.ini` file with consideration for future ET5 support, but those may not be too important.

It's possible that you have to name the file `ET4.bin` for it to flash from the SD card, but I haven't tried. Maybe it can be flashed over USB. I'll mess with it soon and figure out what's required. I'll start out by flashing the latest firmware binary from Anet so I can see a working result first.

thisiskeithb commented:
> The main question is whether the IS_ANET_ET customizations here (making the user interface blue and orange, etc.) are desirable or necessary.

Probably not. That can always be set by the various ET* configs.

> Also, the pins file I have specifies a CUSTOM_CABLE for the SD connection so we might need to look at that.

I saw that in the pins file originally, but didn't see a matching `#ifdef CUSTOM_CABLE` where that'd be useful anywhere, so I removed it. 🤷‍♂️

> It's possible that you have to name the file ET4.bin for it to flash from the SD card, but I haven't tried. Maybe it can be flashed over USB.

I pinged Anet again for more info/a status on where they were at with their own bootloader, but haven't heard back.

Sebazzz commented:
I don't see the build offset of `0x10000`, which is required if you use OpenBlt bootloader (and probably any bootloader). Is that deliberate?

thisiskeithb commented:
The [bootloader I referenced above](https://www.cnx-software.com/2020/02/04/how-to-recover-from-a-bad-firmware-upgrade-on-anet-et4-3d-printer/) doesn't use that offset, so it will likely vary depending on which bootloader you go with.

ST-Link Utility screenshot for the referenced bootloader:
![image](https://user-images.githubusercontent.com/13375512/100527837-25d71580-318b-11eb-956a-a64ea74b1b6a.png)

Sebazzz commented:
> 
> 
> The [bootloader I referenced above](https://www.cnx-software.com/2020/02/04/how-to-recover-from-a-bad-firmware-upgrade-on-anet-et4-3d-printer/) doesn't use that offset, so it will likely vary depending on which bootloader you go with.

That's not a bootloader right? That's flashing directly via ST-Link. OpenBLT allows flashing through SD card. You only need to flash OpenBLT once.

> Anet is working on their own bootloader, so that will also likely change once available.

It will - but we're not there yet 😄

thisiskeithb commented:
> That's not a bootloader right?

Oops, yes.

I now see the following in David's fork below the ET4 environment:

```ini
# Uncomment next line to build for use with BootLoader.
# Offset depends on BL used.
#board_build.offset  = 0x10000

# Uncomment next 2 lines to build for debug.
#build_type = debug
#debug_build_flags = -O0 -ggdb3 -g3
```

So it still depends on how you plan to update. Is OpenBLT the prefered/most common bootloader for Anet ET4/5 modders?

thinkyhead commented:
I have to work with OpenBLT for a project and would enjoy hearing about others' experience configuring this bootloader. I hear it's quite configurable, but the "conf.h" files are pretty sparse.

Sebazzz commented:
Let's just say that taking your printer apart and flashing it with an ST-Link every time you've compiled Marlin gets a bit annoying 😉 

The default precompiled release of OpenBLT works great:
- It logs the flashing log to the SD card 
- It shows progress on the TFT
- It is fast

This is on the Anet ET4 Pro. But since it has the same motherboard I assume this is included in your PR. It just has a little different config.

thisiskeithb commented:
> Let's just say that taking your printer apart and flashing it with an ST-Link every time you've compiled Marlin gets a bit annoying

I get it. Some of my printers with an easy to access SWD header are still a pain to fix with an ST-Link when I need to recover the board. I was just wondering if OpenBLT is *the standard* when it comes to installing a bootloader on these boards since I’m not super familiar with the Anet ET modding scene yet.

If it is, then it makes sense to enable the 0x10000 offset and update the [ET4 readme](https://github.com/MarlinFirmware/Configurations/tree/import-2.0.x/config/examples/Anet/ET4) to call out flashing OpenBLT and include a similar readme with the other pending configs.

> This is on the Anet ET4 Pro. But since it has the same motherboard I assume this is included in your PR.

This PR adds the Anet ET4 *board* which is used in ET4 and ET5 printers. https://github.com/MarlinFirmware/Configurations/pull/322 has configs for more ET4 printers and I hope to add ET5 configs once mine gets here this week & I can verify everything works.

I requested more info from Anet about the board revisions and if any other models use the ET4 board as well as TFT types so we can improve the existing ET4 configs and pending ET configs as-needed.

Sebazzz commented:
> I was just wondering if OpenBLT is _the standard_ when it comes to installing a bootloader on these boards since I’m not super familiar with the Anet ET modding scene yet.

I do believe David's Marlin for is the standard when it comes to the Anet ET series, so I think his bootloader is too.

thisiskeithb committed:
For use with davidtgbe's OpenBLT bootloader

https://github.com/davidtgbe/openblt

thisiskeithb commented:
I just updated the environment to specify using davidtgbe's OpenBLT bootloader and updated `board_build.offset` to  `0x10000`.

I'll be sure to update any related ET series' config readmes to make sure it's clear that ET owners will need to use this bootloader when compiling with this environment.

thisiskeithb commented:
@thinkyhead: Which board/TFT display does your printer have? Here's the "ET4-MB_V1.1" pinout which uses SPI and not FSMC like what was added in those TFT patches:
![image](https://user-images.githubusercontent.com/13375512/101110782-9668a280-358e-11eb-94a7-ff5590344976.png)

(H/T to @rhapsodyv for finding that)

rhapsodyv commented:
I will push the code for TFT SPI on a F4 board, and we can revert some of those commits that changed the FSMC and TFT code.

thisiskeithb commented:
My ET5 X arrived this week, so I will be unboxing it this weekend & getting things setup for testing 🙂

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/20280

Sebazzz commented:
> 
> 
> @thinkyhead: Which board/TFT display does your printer have? Here's the "ET4-MB_V1.1" pinout which uses SPI and not FSMC like what was added in those TFT patches:
> ![image](https://user-images.githubusercontent.com/13375512/101110782-9668a280-358e-11eb-94a7-ff5590344976.png)
> 
> (H/T to @rhapsodyv for finding that)

The ET4 Pro [has a slightly different board as well](https://damsteen.nl/blog/2020/11/30/anet-et4-pro-review#electrical--inside-the-base) marked `ET4P-MB_V1.1`:

![image](https://user-images.githubusercontent.com/1426097/101143679-834de500-3617-11eb-99f9-29f5854fcb79.png)

thisiskeithb commented:
> The ET4 Pro has a slightly different board as well marked ET4P-MB_V1.1

Is that because the Pro models have 2208s instead of A4988s?

I’m still digging around the web/YouTube/Facebook to find as much info as I can about this board and printers as well as the TFTs since Anet hasn’t responded to my inquires (which is odd since they reached out to me first!).

Sebazzz commented:
> Is that because the Pro models have 2208s instead of A4988s?

Yes, they do. In the linked article you can also find a pic of the TFT. 

If you need any info, I have this thing here for the next week - after that it will be going away to a friend of mine.

thisiskeithb commented:
> But I compile it using the same pins file, so I think in that sense it is 100% compatible

Thanks for confirming what I was thinking. We can simply declare the stepper driver and enable other features like the probe via config.

I imagine the 2208s are wired in standalone mode since the existing pins file doesn’t have any TMC/UART-related pins.

Sebazzz commented:
> I imagine the 2208s are wired in standalone mode since the existing pins file doesn’t have any TMC/UART-related pins.

I have my doubts about that. This is my reasoning: On Creality boards, which also use standalone drivers, there are little SMD mounted pots to change the vref. On the ET4 board no such thing exists - so it may be wired in UART. I assume UART wired TMC stepper drivers can still be run in standalone mode too without wiring changes (because you still need to give the X_STEP_PIN etc).

> Thanks for confirming what I was thinking. We can simply declare the stepper driver and enable other features like the probe via config.

Yes, these are the changes I did to make it working well for me: https://github.com/Sebazzz/ANET-ET4Pro-Marlin/commits/easyconfig

thisiskeithb commented:
I just pinged Anet _again_ for schematics and additional information or original firmware to help out, so we'll see if they get back to me quickly.

thinkyhead commented:
AFAIK the ET4 here is the basic unit. No special markings on the box. I'll turn it over and see what kind of electronics it actually has inside in the morning. It's a touch screen with a custom interface, which I can also demo for you then. A very nice looking and solid machine, but I do believe E and Z might not be silent drivers…. But don't let me start rumors until I have a closer look.

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/20280

thinkyhead committed:
Get closer to Sebazzz branch

thisiskeithb commented on Marlin/src/pins/stm32f4/pins_ANET_ET4.h:
```diff
@@ -199,7 +174,7 @@
 
 #ifndef SDCARD_CONNECTION
-  #define SDCARD_CONNECTION              ONBOARD
+  #define SDCARD_CONNECTION         CUSTOM_CABLE
```
This define doesn't do anything since there's no `CUSTOM_CABLE` setup in the pins file. But there's also no `ONBOARD` either.

thinkyhead replied:
Yes, it's mysterious. This is what's in the other fork, so if it doesn't matter it can be left as is. If it might matter in future, we can change to a better value.

thisiskeithb commented on Marlin/src/pins/stm32f4/pins_ANET_ET4.h:
```diff
@@ -128,62 +136,29 @@
+ * Status: Working. Merged implementation from jmz52 fork.
+ * Hardware: TOUCH: XPT2046 => https://ldm-systems.ru/f/doc/catalog/HY-TFT-2,8/XPT2046.pdf
+ */
 
```
Since JMZ52's fork was already merged into Marlin, I originally stripped this comment out.

thisiskeithb commented on Marlin/src/pins/stm32f4/pins_ANET_ET4.h:
```diff
@@ -33,6 +33,12 @@
+ * Status: Working.
+ * Hardware: AT24C04C (ATMLH744 04CM) 4 Kb => http://ww1.microchip.com/downloads/en/DeviceDoc/AT24C04C-AT24C08C-I2C-Compatible-%20Two-Wire-Serial-EEPROM-4-Kbit-8-Kbit-20006127A.pdf
+ */
+
```
Do we really need a comment to say that EEPROM is working? It's defined & setup below, so that's why I stripped it out.

thinkyhead replied:
It's nice to know someone tested it. I haven't yet. Now that we know we can trust the author we can remove these comments.

thisiskeithb commented:
The ET5 X is unboxed!

I opened the case and I have the ET4-MB_V1.1 motherboard and LCD35-ET5_V1.0 TFT. Also, @rhapsodyv has some STM32F4-related/SPI TFT code that I will be testing this weekend, so I'll report back.

| Motherboard | TFT |
| :------------: | :---: |
| <img src="https://user-images.githubusercontent.com/13375512/101227649-777d1580-364d-11eb-8ee9-b49686044773.png"> | <img src="https://user-images.githubusercontent.com/13375512/101227689-9c718880-364d-11eb-851f-d2e507bf26ed.png"> |

thinkyhead committed:
Fewer comments

rhapsodyv commented:
Anyone have the schematics? So we can find the right SPI pins for the TFT.

rhapsodyv committed:
Rollback FSMC changes

rhapsodyv committed:
More FSMC Rollback

thisiskeithb commented:
> Anyone have the schematics? So we can find the right SPI pins for the TFT.

I’ve asked my contact at Anet for more info again, but they haven’t responded yet.

rhapsodyv commented:
I'm almost sure the pins for touch are wrong, because the board photo shows that touch share SPI with TFT, but the current PINS don't belong for any SPI on STM32F407VGT6...

We need the schematics... or trial and error...

rhapsodyv commented:
I reverted the FSCM changes. And I'm almost sure we need to revert the `ENDIAN_COLOR` changes too, but we need to wait for the schematics to be sure.

rhapsodyv committed:
ANET_ET4_TFT28 and ANET_ET5_TFT35 support + FSMC 8bit for SMT32

rhapsodyv committed:
ANET_ET4_TFT28 and ANET_ET5_TFT35 support + FSMC 8bit for SMT32

thisiskeithb commented:
I've flashed OpenBLT and now @rhapsodyv and I are working through the TFT issues. We'll provide updates as we test out the code changes 🙂

thisiskeithb committed:
Add BLTouch warning

thisiskeithb committed:
Hard code ANET_ET5_TFT35 calibration values

TOUCH_SCREEN_CALIBRATION generates wildly inaccurate values, so hardcode known-good/tested values.

thisiskeithb committed:
Hard code ANET_ET4_TFT28 calibration values too

rhapsodyv committed:
Allow touch calibration values on PINs file again

thisiskeithb committed:
Use ET4_TFT28 calibration values from davidtgbe/Marlin

https://github.com/davidtgbe/Marlin/blob/dd1afe86809efb8fe82bb541fd9393c81a26c175/Marlin/Configuration.h#L2339-L2344

thisiskeithb commented:
@rhapsodyv's last commit allows pins to override calibration defaults again so they've been hard coded for `ANET_ET5_TFT35` and `ANET_ET4_TFT28` due to `TOUCH_SCREEN_CALIBRATION` generating some wild values that do not allow touch to work correctly.

@thinkyhead The `ANET_ET5_TFT35` values were tested on my machine, but the `ANET_ET4_TFT28` values were grabbed from [davidtgbe/Marlin](https://github.com/davidtgbe/Marlin/blob/dd1afe86809efb8fe82bb541fd9393c81a26c175/Marlin/Configuration.h#L2339-L2344). Are you able to test these on your machine (if you've flashed the OpenBLT bootloader)?

I think it's safe to say that Marlin works on Anet ET 4/5 printers with OpenBLT as the bootloader 🙂

The existing ET4 config in the Configurations repo will need to be tweaked as well as the pending ET4 configs. I'll have an ET5 X config ready once I play with the printer a bit more this week.

Also, I'll work with @rhapsodyv to see if there's any possibility of signing firmware to use the stock bootloader, but I wouldn't let that hold up the PR.

thisiskeithb committed:
Update motherboard name based off silkscreen

thisiskeithb committed:
Update motherboard name in platformio.ini too

thisiskeithb committed:
Revert baudrate change

thisiskeithb committed:
Merge remote-tracking branch 'MarlinFirmware/bugfix-2.0.x' into pr/add_anet_et4_support

thisiskeithb committed:
Add Anet ET4P motherboard

The ET4P uses TMC2208_STANDALONE drivers and this will now show the proper board name in the About Printer section.

thisiskeithb commented on Marlin/src/pins/stm32f4/pins_ANET_ET4P.h:
```diff
@@ -0,0 +1,34 @@
+#if !AXIS_DRIVER_TYPE_X(TMC2208_STANDALONE) || !AXIS_DRIVER_TYPE_Y(TMC2208_STANDALONE) || !AXIS_DRIVER_TYPE_Z(TMC2208_STANDALONE) || !AXIS_DRIVER_TYPE_E0(TMC2208_STANDALONE)
+  #error "You must set ([XYZ]|E0)_DRIVER_TYPE to TMC2208_STANDALONE in Configuration.h for ET4P-MB_V1.x."
+#endif
+
```
This check was stolen from the Einsy Rambo pins file to ensure users select the proper driver type when using the "Pro" board which has TMC2208s in standalone mode instead of A4988s.

thisiskeithb committed:
Merge remote-tracking branch 'MarlinFirmware/bugfix-2.0.x' into pr/add_anet_et4_support

thisiskeithb committed:
There is no ET5 motherboard

The Anet ET5 series uses the Anet ET4 or Anet ET4P motherboard (yes, naming is confusing).

thisiskeithb commented:
Just to clarify, the Anet ET4 and ET5 series share the same "ET4" motherboards. 

Yes, the naming convention is bad.

The motherboards are labeled ET4 (A4988s) or ET4P (TMC2208s) depending on which printer variant you have.

Sebazzz commented:
To be clear: what is the issue with the TFT? It is working fine for me on my Marlin build.

thisiskeithb commented:
> To be clear: what is the issue with the TFT? It is working fine for me on my Marlin build.

With this PR?

The Anet 2.8" and 3.5" TFTs work correctly with `ANET_ET4_TFT28` and `ANET_ET4_TFT35` defines in the config from this PR.

The only issue I ran into was that `TOUCH_SCREEN_CALIBRATION` would generate some wild values on the Anet TFT35 and did not work correctly, so those values are set via the ET4 motherboard pins file.

thisiskeithb committed:
Merge remote-tracking branch 'MarlinFirmware/bugfix-2.0.x' into pr/add_anet_et4_support

thisiskeithb committed:
Merge remote-tracking branch 'MarlinFirmware/bugfix-2.0.x' into pr/add_anet_et4_support

rhapsodyv committed:
Fix compiling warning, keep interface standard

sjasonsmith commented on buildroot/share/PlatformIO/ldscripts/anet_et4.ld:
```diff
@@ -0,0 +1,184 @@
+/*
+*****************************************************************************
+**
+**  File        : LinkerScript.ld
```
This looks like a duplicate file which should be deleted.

sjasonsmith replied:
It is already in the variant folder, which is where it should be. It seems like mostly STM32F1 boards that have scripts here.

Most of the other boards in this HAL have them only in the variant folder.

thisiskeithb replied:
This isn't the same as [Marlin/buildroot/share/PlatformIO/variants/ANET_ET4/ldscript.ld](https://github.com/MarlinFirmware/Marlin/blob/4986377dd3d53b4990b170bc1e0f112561ec3ebb/buildroot/share/PlatformIO/variants/ANET_ET4/ldscript.ld) and removing it causes the build to fail.

rhapsodyv replied:
Or the two ldscript are equal, or one of them is wrong. Keeping both will lead to confusion. The script `buildroot/share/PlatformIO/scripts/anet_et4.py` is forcing to use this file.
But I think we should "join" this in favor of the variant version (even if we need to fix the variant file).

I think we should try to use our generic scripts too, instead of specific ones, for the new STM32 HAL. So we avoid having one script per board...

sjasonsmith commented on buildroot/share/PlatformIO/boards/ANET_ET4.json:
```diff
@@ -0,0 +1,46 @@
+      ]
+    ],
+    "mcu": "stm32f407vgt6",
+    "variant": "ANET_ET4"
```
I prefer to prefix our board and variant names with MARLIN. You can see that has been done for a few other boards, at least for the variants.

The variants actually get copied into the framework folder, so that makes it obvious which came from the frame work and which Marlin copied into there.

sjasonsmith reviewed (commented):
How would you feel about splitting this into two PRs?

1. New Variant + pins (minus display settings)
    This could probably merge with fairly little effort, while the next PR is worked on.

2. Add TFT changes necessary for it to work.
    I see these as the larger and possibly more controversial part of the PR. Splitting it out should help focus the PR on only those files, without all the distractions from the first PR.

sjasonsmith commented:
Feel free to ignore my "split the PR" if you already have plans with others how to move this forward.

thisiskeithb commented:
> Feel free to ignore my "split the PR" if you already have plans with others how to move this forward.

I’m just waiting on @thinkyhead to confirm everything works on his ET4 printer and then this can be merged.

I have an ET4P motherboard and TFT28 on the way and can test these changes if he doesn’t have time.

thinkyhead committed:
Adjustments

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/20280

thinkyhead commented:
When installing from SD card is there any special requirement for the firmware BIN name? Some revision of this was including `board_build.firmware = et4.bin` and I've had an "`et4.bin`" file lying around that I might have downloaded from Anet, but I'm not sure.

thinkyhead committed:
Tweak pio.ini spacing

thisiskeithb commented:
> When installing from SD card is there any special requirement for the firmware BIN name? Some revision of this was including `board_build.firmware = et4.bin` and I've had an "`et4.bin`" file lying around that I might have downloaded from Anet, but I'm not sure.

For the stock bootloader and stock Anet firmware, it should be `et4.bin`, `et4p.bin`, `et4x.bin`, etc. or `et5.bin`, `et5p.bin`, `et5x.bin`, etc.

When using David’s OpenBLT fork for flashing Marlin from this PR, it should be the generated `firmware.srec` file.

Anet finally responded to me tonight and said that several engineers/developers were unsuccessful in getting Marlin to update with the stock bootloader so you will still need to flash OpenBLT initially with an ST-Link/J-Link in order to update Marlin via SD card on these printers.

thinkyhead committed:
Fewer tft_fsmc files

thisiskeithb commented:
I just submitted https://github.com/MarlinFirmware/Configurations/pull/367 to add ET5 series configs as well as https://github.com/MarlinFirmware/Configurations/pull/368 which corrects some minor board/TFT defines for the ET4 configs (and adds the Anet "Black" theme).

thisiskeithb committed:
Point to the Anet OpenBLT releases & update env comment

rhapsodyv commented on Marlin/src/lcd/tft/tft.h:
```diff
@@ -30,6 +30,12 @@
 #include "../../inc/MarlinConfig.h"
 
+#if IS_ANET_ET
+  #define ENDIAN_COLOR(C) (((C) >> 8) | ((C) << 8))
```
This ENDIAN_COLOR is needed by any 8 bit interface, in the TFT code. It's not a ANET thing.

thisiskeithb commented on Marlin/src/HAL/STM32/tft/tft_fsmc.cpp:
```diff
@@ -74,8 +74,8 @@ void TFT_FSMC::Init() {
-  Timing.DataLatency = 17;
+  Timing.CLKDivision = TERN(TFT_INTERFACE_FSMC_8BIT, 0, 16);
+  Timing.DataLatency = TERN(TFT_INTERFACE_FSMC_8BIT, 0, 17);
   Timing.AccessMode = FSMC_ACCESS_MODE_A;
```
@thinkyhead: @rhapsodyv had me revert this change and there was no difference in how the TFT35 responded. It felt just as snappy with either change, but he said that 0 is an invalid value.

Does your TFT28 act any differently between these changes?

rhapsodyv commented on Marlin/src/HAL/STM32/tft/tft_fsmc.h:
```diff
@@ -121,7 +126,11 @@ const PinMap PinMap_FSMC_CS[] = {
 
-#define FSMC_RS(A)  (void *)((2 << A) - 2)
+#if ENABLED(TFT_INTERFACE_FSMC_8BIT)
+  #define FSMC_RS(A)  (void *)((2 << (A-1)) - 1)
```
I confirmed that we need this. The address of the SRAM is `0x60000000U` + an offset based in the RS pin used. For PD_7, for example, is 0x20000 = 2^16. But, using 16 bit, it's 0x10000.
So, this code is right for 8 bit.

rhapsodyv commented on Marlin/src/HAL/STM32/tft/tft_fsmc.h:
```diff
@@ -98,14 +101,16 @@ const PinMap PinMap_FSMC[] = {
-  {PD_8,   FSMC_NORSRAM_DEVICE, FSMC_PIN_DATA}, // FSMC_D13
-  {PD_9,   FSMC_NORSRAM_DEVICE, FSMC_PIN_DATA}, // FSMC_D14
-  {PD_10,  FSMC_NORSRAM_DEVICE, FSMC_PIN_DATA}, // FSMC_D15
+  #if DISABLED(TFT_INTERFACE_FSMC_8BIT)
```
This code is needed for 8 bit fsmc too. Without this, we would mess up with pins used in another places.

rhapsodyv commented on buildroot/share/PlatformIO/boards/ANET_ET4.json:
```diff
@@ -0,0 +1,46 @@
+  "name": "STM32F407VG (192k RAM. 1024k Flash)",
+  "upload": {
+    "disable_flushing": false,
+    "maximum_ram_size": 196608,
```
hum... this board have 128kb (131072) of ram + 6kb of ccmram... It's only working with `196608` because someone had edited the ldscript...

thisiskeithb replied:
It probably makes sense to revert it to the standard settings.

rhapsodyv commented on buildroot/share/PlatformIO/variants/MARLIN_ANET_ET4/variant.h:
```diff
@@ -0,0 +1,385 @@
+
+// SD card slot Definitions
+// SD detect signal can be defined if required
+#define SD_DETECT_PIN           PD3
```
the variant is the right place for it?

thisiskeithb replied:
Probably not since it's defined in the board's pins file too.

thisiskeithb replied:
Since this file is going away, this will be moot.

rhapsodyv commented on platformio.ini:
```diff
@@ -1228,6 +1228,28 @@ build_flags       = ${common_stm32.build_flags}
+build_unflags      = ${common_stm32.build_unflags}
+  -DUSBCON -DUSBD_USE_CDC -DUSBD_VID=0x0483
+extra_scripts      = ${common.extra_scripts}
+  pre:buildroot/share/PlatformIO/scripts/copy_marlin_variant_to_framework.py
```
maybe we should use our new generic script for variants `pre:buildroot/share/PlatformIO/scripts/generic_create_variant.py`.

thisiskeithb replied:
Let's! Per our discussion & testing via Discord, this seems to work fine. It'll be nice to clean up/dump all those extra files!

rhapsodyv commented on buildroot/share/PlatformIO/variants/MARLIN_ANET_ET4/hal_conf_extra.h:
```diff
@@ -0,0 +1,55 @@
+#define HAL_CORTEX_MODULE_ENABLED
+//#define HAL_UART_MODULE_ENABLED // by default
+//#define HAL_PCD_MODULE_ENABLED  // Since STM32 v3.10700.191028 this is automatically added if any type of USB is enabled (as in Arduino IDE)
+#define HAL_SD_MODULE_ENABLED
```
this board uses SDIO or SD on SPI?

thisiskeithb replied:
Based on our discussion on Discord, SDIO.

thisiskeithb committed:
Merge remote-tracking branch 'MarlinFirmware/bugfix-2.0.x' into pr/add_anet_et4_support

thisiskeithb commented:
I marked this PR as draft again while @rhapsodyv converts this board to use the new generic variant settings.

rhapsodyv committed:
those values work on both 8 and 16 bit

rhapsodyv committed:
we dont need custom variant for anet et4, use generic instead

thisiskeithb commented:
@rhapsodyv works fast 🙂 I just flashed the latest changes and heaters, fans, motors, probe, etc. all work as expected.

rhapsodyv committed:
the endian byte inversion is for all 8 bit interfaces (currently, only fsmc 8 bit)

rhapsodyv reviewed (approved):
I think everything is OK now.

thisiskeithb commented:
@xC0000005: If you're on Marlin's discord, can you reach out to me about reverse engineering the stock bootloader/update process on these boards? @rhapsodyv is confident that you can work your magic like you did for the chitu bootloader mess 🙂

xC0000005 commented on platformio.ini:
```diff
@@ -1241,6 +1241,29 @@ build_flags       = ${common_stm32.build_flags}
+board_build.core     = stm32
+board_build.variant  = MARLIN_F4x7Vx
+board_build.ldscript = ldscript.ld
+board_build.firmware = firmware.srec
```
If your end goal is SD based flashing, this will be named "ET4.BIN"

xC0000005 replied:
No idea if it's ET5.bin on the ET5 machines, but if so, you can take an approach similar to ledge where all of this is an an ANET_ETX_BASE and the different environments override the firmware name.

thisiskeithb replied:
> If your end goal is SD based flashing, this will be named "ET4.BIN"

It will be, but this will use OpenBLT for now since I knew using the stock bootloader wouldn't be possible, or at least not easily.

thisiskeithb replied:
> No idea if it's ET5.bin on the ET5 machines

It depends on the printer series/model. Example: `et4.bin`, `et4p.bin`, `et4x.bin`, `et5.bin`, `et5p.bin`, `et5x.bin`.

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/20280

xC0000005 commented:
If anyone else has one of these and wants to try disk based update, the thing that stops Marlin from ever booting is that the stock boot loader wants the _estack value to be less than 64k (I'd set it to 32k just to be safe), so set maximum_ram_size to ~63k and see if it doesn't work better with completely stock boot loader.

thisiskeithb commented:
I just found out the [Labists ET4](https://labists.com/products/3d-printer-et4-diy-kit) is OEM'd by Anet & uses an ET4 motherboard & TFT28, so this PR works for those printers too! I'll get a config submitted for it soon.

chepo92 commented:
How to wire an STlink V2 to upload the Bootlader? I see the V (or U?) and G marks on a header, but im missing labels of the other 2 pins

thisiskeithb commented:
> How to wire an STlink V2 to upload the Bootlader?

I put together readmes for each of the ET4/5 configs that walk you through everything. Example: https://github.com/MarlinFirmware/Configurations/tree/import-2.0.x/config/examples/Anet/ET5X

chepo92 commented:
Can anyone confirm that this is working with ET4 version with (inductive/capacitive) probe? 
With M119 i see no change in probe status (checked that probe works, lights on  and the signal cable changes voltage when put close to bed) 
Is PC3 pin correct?
#define Z_MIN_PROBE_PIN                   PC3

Sebazzz commented:
@chepo92 Compare with my old Marlin fork: https://github.com/Sebazzz/ANET-ET4Pro-Marlin/
This one does work on the ET4 Pro.
